pkg/utils/k8sutil.go

```diff
@@ -265,247 +268,98 @@ func GetFunctionData(runtime, ftype, modName string) (imageName, depName, fileNa
+	Dependency string
+	Labels     map[string]string
+}
+
 // EnsureK8sResources creates/updates k8s objects (deploy, svc, configmap) for the function
 func EnsureK8sResources(ns, name string, funcObj *spec.Function, client kubernetes.Interface) error {
 	str := strings.Split(funcObj.Spec.Handler, ".")
 	if len(str) != 2 {
 		return errors.New("Failed: incorrect handler format. It should be module_name.handler_name")
 	}
+
 	funcHandler := str[1]
 	modName := str[0]
+
 	fileName := modName
 	imageName, depName, fileName, err := GetFunctionData(funcObj.Spec.Runtime, funcObj.Spec.Type, modName)
-
 	if err != nil {
 		return err
 	}
 
-	//add configmap
 	labels := map[string]string{
 		"function": name,
 	}
 	for k, v := range funcObj.Metadata.Labels {
 		labels[k] = v
 	}
 
+	funcMeta := metadata{
+		Name:       name,
+		Namespace:  ns,
+		Handler:    funcHandler,
+		Module:     modName,
+		Image:      imageName,
+		File:       fileName,
+		Dependency: depName,
+		Labels:     labels,
+	}
+
 	t := true
 	or := []metav1.OwnerReference{
 		{
 			Kind:               "Function",
 			APIVersion:         "k8s.io",
-			Name:               name,
+			Name:               funcMeta.Name,
 			UID:                funcObj.Metadata.UID,
 			BlockOwnerDeletion: &t,
 		},
 	}
 
-	podAnnotations := map[string]string{
-		// Attempt to attract the attention of prometheus.
-		// For runtimes that don't support /metrics,
-		// prometheus will get a 404 and mostly silently
-		// ignore the pod (still displayed in the list of
-		// "targets")
-		"prometheus.io/scrape": "true",
-		"prometheus.io/path":   "/metrics",
-		"prometheus.io/port":   "8080",
-	}
-
-	configMap := &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            name,
-			Labels:          labels,
-			OwnerReferences: or,
-		},
-		Data: map[string]string{
-			"handler": funcObj.Spec.Handler,
-			fileName:  funcObj.Spec.Function,
-			depName:   funcObj.Spec.Deps,
-		},
-	}
-
-	_, err = client.Core().ConfigMaps(ns).Create(configMap)
-	if err != nil && k8sErrors.IsAlreadyExists(err) {
-		data, _ := json.Marshal(configMap)
-		_, err = client.Core().ConfigMaps(ns).Patch(configMap.Name, types.StrategicMergePatchType, data)
-	}
+	err = ensureFuncConfigMap(client, funcMeta, funcObj, or)
 	if err != nil {
 		return err
 	}
 
-	//add service
-	svc := &v1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            name,
-			Labels:          labels,
-			OwnerReferences: or,
-		},
-		Spec: v1.ServiceSpec{
-			Ports: []v1.ServicePort{
-				{
-					Port:       8080,
-					TargetPort: intstr.FromInt(8080),
-					Protocol:   v1.ProtocolTCP,
-				},
-			},
-			Selector: labels,
-			Type:     v1.ServiceTypeClusterIP,
-		},
-	}
-	_, err = client.Core().Services(ns).Create(svc)
-	if err != nil && k8sErrors.IsAlreadyExists(err) {
-		data, _ := json.Marshal(svc)
-		_, err = client.Core().Services(ns).Patch(svc.Name, types.StrategicMergePatchType, data)
-
-	}
+	err = ensureFuncService(client, funcMeta, or)
```

Comment:
Ditto - pass the full `Function` object, not ad-hoc imitations.  If you want to avoid duplicating the code that creates the ownerRef, then create a function that creates the ownerRef from Function object, and call it from multiple places.